SLI sleep implementation conforming to standards?

@heplesser requested to create an issue:
It is rather surprising that no-one else has run into this issue:
```
[ 16%] Building CXX object sli/CMakeFiles/sli_lib.dir/slidata.cc.o
/Users/ehagen/Repositories/nest-simulator/sli/slicontrol.cc:1725:30: error: non-constant-expression cannot be narrowed from type 'long' to '__darwin_suseconds_t' (aka 'int') in
   initializer list [-Wc++11-narrowing]
struct timeval tv = { sec, usec };
                          ^~~~
```
In Sleep_dFunction::execute( SLIInterpreter* i ), usec is defined as long, but the POSIX standard only requires an "signed integer type capable of storing values at least in the range [-1, 1000000]", and on OSX it is of type int.

C++98 didn't care about the narrow, C++11 does. The way this SLI-sleep functions is implemented using the POSIX select() function also seems archaic. There is POSIX usleep() and since C++11 even std::this_thread::sleep_for.

Encountered on MacOS with Anaconda Python and conda's mpi compilers:
```
I think I found a new and interesting way to break the compilation of NEST, this time with Anaconda Python 3 (most recent) on my mac (v10.13.5):
$ conda -V
conda 4.5.4
$ python -V
Python 3.6.5

I’ve tried to use dependencies and compilers as provided by Anaconda and conda-forge to keep things self contained;
$ conda config --add channels conda-forge
$ conda install openmpi readline pkg-config gsl doxygen cmake #and probably others

Nest installation procedure (master@ad52581ce7):
$ export CC=~/anaconda3/bin/mpicc
$ export CXX=~/anaconda3/bin/mpicxx

$ export NESTBLD=$HOME/Repositories/nest-simulator-bld
$ export NESTSOURCES=$HOME/Repositories/nest-simulator
$ export NESTPREFIX=$HOME/.local/nest_py36

$ mkdir $NESTBLD
$ cd $NESTBLD

$ cmake -DCMAKE_INSTALL_PREFIX:PATH=$NESTPREFIX \
    -Dwith-mpi=ON \
    -Dwith-python=ON \
    -DLTDL_ROOT_DIR=~/anaconda3 -Dwith-gsl=~/anaconda3 -Dwith-readline=~/anaconda3 \
    -DCMAKE_C_COMPILER=$CC \
    -DCMAKE_CXX_COMPILER=$CXX \
    $NESTSOURCES


Which gives (looks fine to me, except OpenMP is not found - don’t think it is crucial):
-- The CXX compiler identification is Clang 6.0.0
-- The C compiler identification is Clang 6.0.0
-- Check for working CXX compiler: /Users/ehagen/anaconda3/bin/mpicxx
-- Check for working CXX compiler: /Users/ehagen/anaconda3/bin/mpicxx -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /Users/ehagen/anaconda3/bin/mpicc
-- Check for working C compiler: /Users/ehagen/anaconda3/bin/mpicc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found Doxygen: /Users/ehagen/anaconda3/bin/doxygen (found version "1.8.14") found components:  doxygen dot 
-- Looking for include file inttypes.h
-- Looking for include file inttypes.h - found
-- Looking for include file mach-o/dyld.h
-- Looking for include file mach-o/dyld.h - found
-- Looking for include file mach/mach.h
-- Looking for include file mach/mach.h - found
-- Looking for include file memory.h
-- Looking for include file memory.h - found
-- Looking for include file stdint.h
-- Looking for include file stdint.h - found
-- Looking for include file sys/types.h
-- Looking for include file sys/types.h - found
-- Looking for C++ include istream
-- Looking for C++ include istream - found
-- Looking for C++ include ostream
-- Looking for C++ include ostream - found
-- Looking for C++ include sstream
-- Looking for C++ include sstream - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of long long
-- Check size of long long - done
-- Check size of u_int16_t
-- Check size of u_int16_t - failed
-- Check size of uint16_t
-- Check size of uint16_t - done
-- Looking for NAN
-- Looking for NAN - found
-- Looking for isnan
-- Looking for isnan - found
-- Looking for M_E
-- Looking for M_E - found
-- Looking for M_PI
-- Looking for M_PI - found
-- Looking for expm1
-- Looking for expm1 - found
-- Host triple: x86_64-apple-darwin
-- Target triple: x86_64-apple-darwin
-- Found LTDL: /Users/ehagen/anaconda3/lib/libltdl.dylib (found version "2.4.6") 
-- Found Readline: /Users/ehagen/anaconda3/lib/libreadline.dylib (found version "7.0") 
-- Found PkgConfig: /Users/ehagen/anaconda3/bin/pkg-config (found version "0.29.2") 
-- Found GSL: /Users/ehagen/anaconda3/include (found version "2.1") 
-- Found PythonInterp: /Users/ehagen/anaconda3/bin/python (found version "3.6.5") 
-- Found PythonLibs: /Users/ehagen/anaconda3/lib/libpython3.6m.dylib (found suitable exact version "3.6.5") 
-- Found Cython: /Users/ehagen/anaconda3/bin/cython (found version "0.28.3") 
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND) 
-- Found MPI_C: /Users/ehagen/anaconda3/bin/mpicc (found version "3.1") 
-- Found MPI_CXX: /Users/ehagen/anaconda3/bin/mpicxx (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- Check the abort exitcode.
-- Check the abort exitcode. 134
-- Check the segmentation fault exitcode.
-- Check the segmentation fault exitcode. 134
-- Check whether the compiler ignores cmath makros.
-- Check whether the compiler ignores cmath makros. OFF
-- Check whether the compiler does NOT include <*.h> headers ISO conformant.
-- Check whether the compiler does NOT include <*.h> headers ISO conformant. OFF
-- Check whether the compiler respects symbolic signal names in signal.h.
-- Check whether the compiler respects symbolic signal names in signal.h. OFF
-- Check static template member declaration.
-- Check static template member declaration. OFF
-- Check for STL vector capacity base unity.
-- Check for STL vector capacity base unity. OFF
-- Check for STL vector capacity doubling strategy.
-- Check for STL vector capacity doubling strategy. OFF
-- Check whether the compiler fails with ICE.
-- Check whether the compiler fails with ICE. OFF
-- Check if ::nan is available from cmath.
-- Check if ::nan is available from cmath. ON
-- Check if ::isnan is available from cmath.
-- Check if ::isnan is available from cmath. ON
ad52581c
-- Found PythonLibs: /Users/ehagen/anaconda3/lib/libpython3.6m.dylib (found version "3.6.5") 

--------------------------------------------------------------------------------
NEST Configuration Summary
--------------------------------------------------------------------------------

Build type          : 
Target System       : Darwin
Cross Compiling     : FALSE
C compiler          : /Users/ehagen/anaconda3/bin/mpicc
C compiler flags    :  -O2 -Wall    
C++ compiler        : /Users/ehagen/anaconda3/bin/mpicxx
C++ compiler flags  :  -O2 -Wall  
Build dynamic       : ON

Built-in modules    : models;precise;topology
User modules        : None
Python bindings     : Yes (Python 3.6: /Users/ehagen/anaconda3/bin/python)
     Includes     : /Users/ehagen/anaconda3/include/python3.6m
     Libraries    : /Users/ehagen/anaconda3/lib/libpython3.6m.dylib

Cython bindings     : Yes (Cython 0.28.3: /Users/ehagen/anaconda3/bin/cython)
Use threading       : No
Use GSL             : Yes (GSL 2.1)
  Includes        : /Users/ehagen/anaconda3/include
  Libraries       : /Users/ehagen/anaconda3/lib/libgsl.dylib;/Users/ehagen/anaconda3/lib/libgslcblas.dylib

Use Readline        : Yes (GNU Readline 7.0)
  Includes        : /Users/ehagen/anaconda3/include
  Libraries       : /Users/ehagen/anaconda3/lib/libreadline.dylib;/Users/ehagen/anaconda3/lib/libncurses.dylib

Use libltdl         : Yes (LTDL 2.4.6)
  Includes        : /Users/ehagen/anaconda3/include
  Libraries       : /Users/ehagen/anaconda3/lib/libltdl.dylib

Use doxygen         : Yes (/Users/ehagen/anaconda3/bin/doxygen)
                  : target `doc` available
  `dot` available : Yes (/opt/local/bin/dot)
                  : target `fulldoc` available
Use MPI             : Yes (MPI: /Users/ehagen/anaconda3/bin/mpicxx)
  FLAGS           : 
  Includes        : 
  Link Flags      : 
  Libraries       : 

Use MUSIC           : No
Use libneurosim     : No

————————————————————————————————————————


But, make fails:
$ make -j4
Scanning dependencies of target nestutil
[  0%] Building CXX object libnestutil/CMakeFiles/nestutil.dir/logging_event.cpp.o
[  0%] Building CXX object libnestutil/CMakeFiles/nestutil.dir/numerics.cpp.o
[  0%] Building CXX object libnestutil/CMakeFiles/nestutil.dir/stopwatch.cpp.o
[  1%] Building CXX object libnestutil/CMakeFiles/nestutil.dir/propagator_stability.cpp.o
[  2%] Linking CXX shared library libnestutil.dylib
[  2%] Built target nestutil
Scanning dependencies of target sli_lib
[  3%] Building CXX object sli/CMakeFiles/sli_lib.dir/booldatum.cc.o
[  3%] Building CXX object sli/CMakeFiles/sli_lib.dir/charcode.cc.o
[  4%] Building CXX object sli/CMakeFiles/sli_lib.dir/arraydatum.cc.o
[  4%] Building CXX object sli/CMakeFiles/sli_lib.dir/allocator.cpp.o
[  5%] Building CXX object sli/CMakeFiles/sli_lib.dir/datum.cc.o
[  5%] Building CXX object sli/CMakeFiles/sli_lib.dir/dict.cc.o
[  5%] Building CXX object sli/CMakeFiles/sli_lib.dir/dictdatum.cc.o
[  5%] Building CXX object sli/CMakeFiles/sli_lib.dir/dictstack.cc.o
[  6%] Building CXX object sli/CMakeFiles/sli_lib.dir/dictutils.cc.o
[  6%] Building CXX object sli/CMakeFiles/sli_lib.dir/doubledatum.cc.o
[  7%] Building CXX object sli/CMakeFiles/sli_lib.dir/fdstream.cc.o
[  7%] Building CXX object sli/CMakeFiles/sli_lib.dir/filesystem.cc.o
[  7%] Building CXX object sli/CMakeFiles/sli_lib.dir/functiondatum.cc.o
[  8%] Building C object sli/CMakeFiles/sli_lib.dir/get_mem.c.o
[  8%] Building CXX object sli/CMakeFiles/sli_lib.dir/integerdatum.cc.o
[  9%] Building CXX object sli/CMakeFiles/sli_lib.dir/interpret.cc.o
[  9%] Building CXX object sli/CMakeFiles/sli_lib.dir/iostreamdatum.cc.o
[  9%] Building CXX object sli/CMakeFiles/sli_lib.dir/iteratordatum.cc.o
[ 10%] Building CXX object sli/CMakeFiles/sli_lib.dir/literaldatum.cc.o
[ 10%] Building CXX object sli/CMakeFiles/sli_lib.dir/name.cc.o
[ 11%] Building CXX object sli/CMakeFiles/sli_lib.dir/slinames.cpp.o
[ 11%] Building CXX object sli/CMakeFiles/sli_lib.dir/namedatum.cc.o
[ 11%] Building CXX object sli/CMakeFiles/sli_lib.dir/oosupport.cc.o
[ 12%] Building CXX object sli/CMakeFiles/sli_lib.dir/parser.cc.o
[ 12%] Building CXX object sli/CMakeFiles/sli_lib.dir/processes.cc.o
[ 13%] Building C object sli/CMakeFiles/sli_lib.dir/psignal.c.o
[ 13%] Building CXX object sli/CMakeFiles/sli_lib.dir/scanner.cc.o
[ 14%] Building CXX object sli/CMakeFiles/sli_lib.dir/sli_io.cc.o
[ 14%] Building CXX object sli/CMakeFiles/sli_lib.dir/sliactions.cc.o
[ 14%] Building CXX object sli/CMakeFiles/sli_lib.dir/sliarray.cc.o
[ 15%] Building CXX object sli/CMakeFiles/sli_lib.dir/slibuiltins.cc.o
[ 15%] Building CXX object sli/CMakeFiles/sli_lib.dir/slicontrol.cc.o
[ 16%] Building CXX object sli/CMakeFiles/sli_lib.dir/slidata.cc.o
/Users/ehagen/Repositories/nest-simulator/sli/slicontrol.cc:1725:30: error: non-constant-expression cannot be narrowed from type 'long' to '__darwin_suseconds_t' (aka 'int') in
    initializer list [-Wc++11-narrowing]
struct timeval tv = { sec, usec };
                           ^~~~
/Users/ehagen/Repositories/nest-simulator/sli/slicontrol.cc:1725:30: note: insert an explicit cast to silence this issue
struct timeval tv = { sec, usec };
                           ^~~~
                           static_cast<__darwin_suseconds_t>( )
[ 16%] Building CXX object sli/CMakeFiles/sli_lib.dir/slidict.cc.o
1 error generated.
make[2]: *** [sli/CMakeFiles/sli_lib.dir/slicontrol.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [sli/CMakeFiles/sli_lib.dir/all] Error 2
make: *** [all] Error 2
```

--- a/lib/sli/sli-init.sli
+++ b/lib/sli/sli-init.sli
@@ -632,9 +632,9 @@ SeeAlso: over, index
 
 /* BeginDocumentation
  Name: sleep - Pauses current process.
- Synopsis:  x sleep -> -
+ Synopsis:  t sleep -> -
  Description: 
-  Pauses the process for x seconds. n can be integer or double.
+  Pauses the process for t seconds. t can be integer or double.
   Resolution is given by pclockspersec.
  Examples:
   5 sleep %wait for 5 seconds
@@ -644,7 +644,7 @@ SeeAlso: over, index
 */
 
 /sleep trie
-  [/integertype] /sleep_i load addtotrie
+  [/integertype] { cvd sleep_d } bind addtotrie
   [/doubletype]  /sleep_d load addtotrie
 def
 

--- a/sli/slicontrol.cc
+++ b/sli/slicontrol.cc
@@ -35,7 +35,9 @@
 
 // Generated includes:
 #include "config.h"
-#include "config.h"
+
+// Includes from libnestutil
+#include "compose.hpp"
 
 // Includes from sli:
 #include "arraydatum.h"
@@ -1680,51 +1682,32 @@ TimeFunction::execute( SLIInterpreter* i ) const
   i->OStack.push_move( tmp );
 }
 
-/*BeginDocumentation
-Name: sleep_i - suspends proces for n seconds
-Synopsis:  n sleep_i -> -
-Description:
-Calls the POSIX select() function.
-SeeAlso: clock, usertime, tic, toc
-*/
-
-void
-Sleep_iFunction::execute( SLIInterpreter* i ) const
-{
-  i->assert_stack_load( 1 );
-
-  const long sec = static_cast< long >( i->OStack.pick( 0 ) );
-  const long usec = 0;
-  struct timeval tv = { sec, usec };
-  if ( sec > 0 )
-  {
-    select( 0, 0, 0, 0, &tv );
-  }
-
-  i->OStack.pop();
-  i->EStack.pop();
-}
-
-/*BeginDocumentation
-Name: sleep_i - suspends proces for x seconds
-Synopsis:  x sleep_d -> -
-Description:
-Calls the POSIX select() function.
-SeeAlso: clock, usertime, tic, toc
-*/
 
 void
 Sleep_dFunction::execute( SLIInterpreter* i ) const
 {
   i->assert_stack_load( 1 );
 
-  const long sec = 0;
-  const long usec = static_cast< long >(
-    static_cast< double >( i->OStack.pick( 0 ) ) * 1000000. );
+  const double t = static_cast< double >( i->OStack.pick( 0 ) );
 
-  struct timeval tv = { sec, usec };
-  if ( usec > 0 )
+  if ( t < 0 )
   {
+    throw BadParameterValue( "t >= 0 required." );
+  }
+  else if ( t > std::numeric_limits< int >::max() )
+  {
+    throw BadParameterValue( String::compose(
+      "t < %1s required.", std::numeric_limits< int >::max() ) );
+  }
+  else if ( t > 0 )
+  {
+    // split into second and microsecond part
+    // limited to 32-bit signed int, see #973.
+    const int t_sec = static_cast< int >( t );
+    const int t_musec = static_cast< int >( ( t - t_sec ) * 1e6 );
+
+    // on some systems, select may modify tv, therefore, it cannot be const
+    struct timeval tv = { t_sec, t_musec };
     select( 0, 0, 0, 0, &tv );
   }
 
@@ -2132,7 +2115,6 @@ const PclockspersecFunction pclockspersecfunction;
 const PgetrusageFunction pgetrusagefunction;
 const TimeFunction timefunction;
 const Sleep_dFunction sleep_dfunction;
-const Sleep_iFunction sleep_ifunction;
 
 const Token_sFunction token_sfunction;
 const Token_isFunction token_isfunction;
@@ -2225,7 +2207,6 @@ init_slicontrol( SLIInterpreter* i )
   i->createcommand( "pgetrusage", &pgetrusagefunction );
   i->createcommand( "time", &timefunction );
   i->createcommand( "sleep_d", &sleep_dfunction );
-  i->createcommand( "sleep_i", &sleep_ifunction );
 
   i->createcommand( "token_s", &token_sfunction );
   i->createcommand( "token_is", &token_isfunction );

--- a/sli/slicontrol.cc
+++ b/sli/slicontrol.cc
@@ -34,6 +34,7 @@
 #include <unistd.h>
 
 // C++ includes:
+#include <algorithm>
 #include <limits>
 
 // Generated includes:
@@ -1704,14 +1705,14 @@ Sleep_dFunction::execute( SLIInterpreter* i ) const
       "t < %1s required.", std::numeric_limits< int >::max() ) );
   }
 
-  // split into second and microsecond part
-  // limited to 32-bit signed int, see #973.
-  const int t_sec = static_cast< int >( t );
-  const int t_musec = static_cast< int >( ( t - t_sec ) * 1e6 );
+  /* Since sleep() only handles entire seconds and usleep() only
+   * times shorter than 1s, we need to split the sleep; see #973. */
+  const unsigned int t_sec = static_cast< unsigned int >( t );
+  const unsigned int t_musec =
+    std::min( 999999U, static_cast< unsigned int >( ( t - t_sec ) * 1e6 ) );
 
-  // on some systems, select may modify tv, therefore, it cannot be const
-  struct timeval tv = { t_sec, t_musec };
-  select( 0, 0, 0, 0, &tv );
+  sleep( t_sec );
+  usleep( t_musec );
 
   i->OStack.pop();
   i->EStack.pop();

--- a/sli/slicontrol.cc
+++ b/sli/slicontrol.cc
@@ -33,6 +33,9 @@
 #include <time.h>
 #include <unistd.h>
 
+// C++ includes:
+#include <limits>
+
 // Generated includes:
 #include "config.h"
 

--- a/sli/slicontrol.cc
+++ b/sli/slicontrol.cc
@@ -1697,22 +1697,21 @@ Sleep_dFunction::execute( SLIInterpreter* i ) const
   {
     throw BadParameterValue( "t >= 0 required." );
   }
-  else if ( t > std::numeric_limits< int >::max() )
+
+  if ( t > std::numeric_limits< int >::max() )
   {
     throw BadParameterValue( String::compose(
       "t < %1s required.", std::numeric_limits< int >::max() ) );
   }
-  else if ( t > 0 )
-  {
-    // split into second and microsecond part
-    // limited to 32-bit signed int, see #973.
-    const int t_sec = static_cast< int >( t );
-    const int t_musec = static_cast< int >( ( t - t_sec ) * 1e6 );
 
-    // on some systems, select may modify tv, therefore, it cannot be const
-    struct timeval tv = { t_sec, t_musec };
-    select( 0, 0, 0, 0, &tv );
-  }
+  // split into second and microsecond part
+  // limited to 32-bit signed int, see #973.
+  const int t_sec = static_cast< int >( t );
+  const int t_musec = static_cast< int >( ( t - t_sec ) * 1e6 );
+
+  // on some systems, select may modify tv, therefore, it cannot be const
+  struct timeval tv = { t_sec, t_musec };
+  select( 0, 0, 0, 0, &tv );
 
   i->OStack.pop();
   i->EStack.pop();

--- a/sli/slicontrol.h
+++ b/sli/slicontrol.h
@@ -535,15 +535,6 @@ public:
   void execute( SLIInterpreter* ) const;
 };
 
-class Sleep_iFunction : public SLIFunction
-{
-public:
-  Sleep_iFunction()
-  {
-  }
-  void execute( SLIInterpreter* ) const;
-};
-
 class Sleep_dFunction : public SLIFunction
 {
 public:

